src/connector/src/sink/snowflake.rs

```diff
@@ -0,0 +1,279 @@
+    s3_client: SnowflakeS3Client,
+    row_encoder: JsonEncoder,
+    row_counter: u32,
+    payload: String,
+    sink_file_suffix: u32,
+}
+
+impl SnowflakeSinkWriter {
+    pub async fn new(
+        config: SnowflakeConfig,
+        schema: Schema,
+        pk_indices: Vec<usize>,
+        is_append_only: bool,
+    ) -> Self {
+        let http_client = SnowflakeHttpClient::new(
+            config.common.account_identifier.clone(),
+            config.common.user.clone(),
+            config.common.database.clone(),
+            config.common.schema.clone(),
+            config.common.pipe.clone(),
+            config.common.rsa_public_key_fp.clone(),
+            config.common.private_key.clone(),
+            HashMap::new(),
+        );
+
+        let s3_client = SnowflakeS3Client::new(
+            config.common.s3_bucket.clone(),
+            config.common.aws_access_key_id.clone(),
+            config.common.aws_secret_access_key.clone(),
+            config.common.aws_region.clone(),
+        )
+        .await;
+
+        Self {
+            config,
+            schema: schema.clone(),
+            pk_indices,
+            is_append_only,
+            http_client,
+            s3_client,
+            row_encoder: JsonEncoder::new(
+                schema,
+                None,
+                super::encoder::DateHandlingMode::String,
+                TimestampHandlingMode::String,
+                TimestamptzHandlingMode::UtcString,
+                TimeHandlingMode::String,
+            ),
+            row_counter: 0,
+            payload: String::new(),
+            // Start from 0, i.e., `RW_SNOWFLAKE_S3_SINK_FILE_0`
+            sink_file_suffix: 0,
+        }
+    }
+
+    /// Note that we shall NOT reset the `sink_file_suffix`
+    /// since we need to incrementally keep the sink
+    /// file *unique*, otherwise snowflake will not
+    /// sink it from external stage (i.e., s3)
+    fn reset(&mut self) {
+        self.payload.clear();
+        self.row_counter = 0;
+    }
+
+    fn at_sink_threshold(&self) -> bool {
+        self.row_counter >= MAX_BATCH_ROW_NUM
+    }
+
+    fn append_only(&mut self, chunk: StreamChunk) -> Result<()> {
+        for (op, row) in chunk.rows() {
+            if op != Op::Insert {
+                continue;
+            }
+            let row_json_string = Value::Object(self.row_encoder.encode(row)?).to_string();
+            self.payload.push_str(&row_json_string);
+            self.row_counter += 1;
+        }
+        Ok(())
+    }
+}
+
+#[async_trait]
+impl SinkWriter for SnowflakeSinkWriter {
+    async fn begin_epoch(&mut self, _epoch: u64) -> Result<()> {
+        Ok(())
+    }
+
+    async fn abort(&mut self) -> Result<()> {
+        Ok(())
+    }
+
+    async fn update_vnode_bitmap(&mut self, _vnode_bitmap: Arc<Bitmap>) -> Result<()> {
+        Ok(())
+    }
+
+    async fn barrier(&mut self, _is_checkpoint: bool) -> Result<Self::CommitMetadata> {
+        Ok(())
```

Comment:
Hmmm, what if a checkpoint barrier comes and there is still some rows in buffer `payload`? IIUC, you must commit these upon checkpoint.

Reply:
Indeed, thanks for pointing this out!

Reply:
Since our ckpt interval is relatively short, it may not meet the `max_batch_row_num` specified by user(s), thus we may need to explicitly mention this in document.

As an example, sink may be triggered by ckpt per ~100 row(s) on average, so even if user set the `max_batch_row_num` greater than `100` (e.g., to `1000`), the actual number of rows being sent to S3 may not fit this threshold.

Reply:
This is exactly the case to be solved by log store aka. sink decouple. May ask @wenym1 for more details.

Reply:
Yes. We can following the similar implementation from iceberg https://github.com/risingwavelabs/risingwave/pull/15634